### PR TITLE
LRDOCS-5037 dynamic includes

### PR DESCRIPTION
**Pull request by mwilliams2014**
@jbalsas this is a start, but I need your help to provide practical examples and review the descriptions of the keys. Right now I have a couple basic screenshots that show where content would be added, but seeing as you can add JS or other resources in the head or body of the theme, the screenshots may not be completely accurate.

### CODE REVIEW

**Comment by jbalsas** [comment]
Hey @mwilliams2014, let's split this into smaller tutorials following what we've been doing with the rest. I find it a bit confusing to see "Editor Extension Points" mixed with "Theme Extension Points". Also, it might be better to call them following their names, same as with the ui tags.

**Comment by mwilliams2014** [comment]
K. I can definitely do that. Do you have working examples that we can show and provide screenshots for? Also, are the descriptions accurate?

**Comment by jbalsas** [comment]
No, descriptions are not accurate, I was going to go over them now :)

**Comment by mwilliams2014** [comment]
Ok. Thanks!

File: develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+Add resources to the TinyMCE editor, such as plugins:
+
+    com.liferay.frontend.editor.tinymce.web#tinymce#additionalResources
+
+Listen to TinyMCE editor events, customize the editor's configuration, etc:
+
+    com.liferay.frontend.editor.tinymce.web#tinymce#onEditorCreate
+
+## Theme Extension Points [](id=theme-extension-points)
```
Comment (jbalsas) [inline develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown]: This title is a bit confusing as well. Unfortunately, we were forced to follow the jsp path name for the keys, which makes it confusing most of the time.

It would be probably simpler to just split it into:
- top_head.jsp
- top_js.jspf
- bottom.jsp

File: develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+top_head:
+https://github.com/liferay/liferay-portal/blob/7.1.x/portal-web/docroot/html/common/themes/top_head.jsp
+
+Load additional resources in the theme's head before the theme's resources are 
+loaded:
+
+    /html/common/themes/top_head.jsp#pre
+
+![Figure 1:](../../images/dynamic-include-top-head-pre.png)
```
Comment (jbalsas) [inline develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown]: This being in the `<head>` tag, it shouldn't add content, just maybe an additional `<link>` tag. You can check out existing cases in portal code.

It might be interesting to show the code snippet of the type of code you are going to inject. It makes it easier to understand.

File: develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+Load additional resources in the theme's head after all theme resources are 
+loaded. Including code here gives you access to the theme's CSS and JavaScript 
+references:
+
+    /html/common/themes/top_head.jsp#post
```
Comment (jbalsas) [inline develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown]: Similar comment here

File: develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+top_js:
+https://github.com/liferay/liferay-portal/blob/7.1.x/portal-web/docroot/html/common/themes/top_js.jspf
```
Comment (jbalsas) [inline develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown]: Not sure about what this link represents

Comment (jbalsas) [reply develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown]: You can check usages of this key in portal as well

File: develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+Include additional JavaScript resources in the theme's head. This is loaded 
+before portlet JavaScript references:
+
+    /html/common/themes/top_js.jspf#resources
+
+bottom:
+https://github.com/liferay/liferay-portal/blob/7.1.x/portal-web/docroot/html/common/themes/bottom.jsp
```
Comment (jbalsas) [inline develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown]: Not sure about this link either

File: develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+Load additional resources in the bottom of the theme's body before the theme's
```
Comment (jbalsas) [inline develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown]: The "resources" mention here is confusing. This will render inside the `body` tag, so it is expected to add maybe some additional html elements:
```
// Example of an additional html element
<div>My additional thing</div> // should make it more beautiful ;)

// Example of a script
<script type="text/javascript">alert('hello');</script>
```

File: develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+resources are loaded in the bottom:
+
+    /html/common/themes/bottom.jsp#pre
+
+![Figure 2:](../../images/dynamic-include-bottom-jsp-pre.png)
+
+Load additional resources in the bottom of the theme's body after the theme's 
+resources are loaded in the bottom. This key gives you access to the theme's 
+portlet CSS references, portlet JavaScript references, and theme JavaScript 
+references. For example, the Simulation Panel is added with this dynamic include:
```
Comment (jbalsas) [inline develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown]: This wording is a bit confusing as well... I think an example is worth 100 words in this case. I would just take inspiration on the Simulation Panel usage and provide a small code snippet showing the example and the result.

File: develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown [hunk 8]
```diff
@@ -0,0 +1,98 @@
+# Dynamic Includes [](id=dynamic-includes)
+
+Dynamic includes, let you inject additional content into apps at key extension 
+points--to add resources, modify editors, etc. Several of these extension points 
+are available for you to use. To use a dynamic include, you must provide the 
+proper key. This tutorial list the available dynamic include keys, along with
```
[the remark this answers is not in the document]
Comment (mwilliams2014) [reply develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown]: Yes, but now this tutorial is going to be broken up into multiple tutorials as you suggested. so "This section of tutorials lists the available dynamic include keys..." is what I'll update it to.

File: develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown (hunk 2 above)
Comment (mwilliams2014) [reply develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown]: Ok, so show the code snippet for this instead of a screenshot?

File: develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown (hunk 4 above)
Comment (mwilliams2014) [reply develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown]: The link is just a reference for me to the JSP that resources is injected into.

File: develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown (hunk 5 above)
Comment (mwilliams2014) [reply develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown]: Same as above

Comment (mwilliams2014) [reply develop/tutorials/articles/120-customizing/10-dynamic-include-extension-points.markdown]: @jbalsas I could only find one usage of this in portal: https://github.com/liferay/liferay-portal/blob/7.1.x/modules/apps/foundation/frontend-js/frontend-js-top-head-extender/src/main/java/com/liferay/frontend/js/top/head/extender/internal/servlet/taglib/TopHeadDynamicInclude.java and it's a bit confusing to follow. Do you have an easier example I can show?

**Comment by mwilliams2014** [comment]
@jbalsas I've updated a few of the dynamic include sections. Please let me know if this is going in the right direction. I've decided to include screenshots of the markup in the developer console to show exactly where the resources are imported.

File: develop/tutorials/articles/120-customizing/10-dynamic-includes/01-wysiwyg-editor-dynamic-includes.markdown [hunk 9]
```diff
@@ -0,0 +1,112 @@
+# WYSIWYG Editor Dynamic Includes
+
+All WYSIWYG editors share the same dynamic include extension points that let you 
+do the following things:
+
+- Add resources, plugins, etc. to the editor:
+    
+    com.liferay.frontend.editor.editorType.web#editorName#additionalResources
```
Comment (jbalsas) [inline develop/tutorials/articles/120-customizing/10-dynamic-includes/01-wysiwyg-editor-dynamic-includes.markdown]: We need to do something to distinguish that these are keys and their placeholder portions, maybe something like

`com.liferay.frontend.editor.%editorType%.web#%editorName%#additionalResources`

or

> com.liferay.frontend.editor.`editorType`.web#`editorName`#additionalResources

File: develop/tutorials/articles/120-customizing/10-dynamic-includes/01-wysiwyg-editor-dynamic-includes.markdown [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+- alloyeditor
+- alloyeditor_bbcode
+- alloyeditor_creole
+- ckeditor
+- ckeditor_bbcode
+- ckeditor_creole
+- simple
+- tinymce
+- tinymce_simple
```
Comment (jbalsas) [inline develop/tutorials/articles/120-customizing/10-dynamic-includes/01-wysiwyg-editor-dynamic-includes.markdown]: This makes more sense as a table, which should also be more readable and clear than a couple of lists

editorType | editorName
--- | ---
alloyeditor | alloyeditor
 | | alloyeditor_bbcode
 | | alloyeditor_creole
ckeditor | ckeditor
 | | ckeditor_bbcode
 | | ckeditor_creole
simple | simple
tinymce | tinymce
 | | tinymce_simple

File: develop/tutorials/articles/120-customizing/10-dynamic-includes/01-wysiwyg-editor-dynamic-includes.markdown [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    	}
+
+    	private BundleContext _bundleContext;
+
+    }
+    
+
+Injected JavaScript (`dialog_definition.js`):
+
+    CKEDITOR.on(
```
Comment (jbalsas) [inline develop/tutorials/articles/120-customizing/10-dynamic-includes/01-wysiwyg-editor-dynamic-includes.markdown]: Here I would better have something like this:

```
// $editorType instance is already available in the execution context
ckEditor.on(
    'paste',
    function(event) {
        event.stop();

        alert('Please, do not paste code here!');
    }
);
```

File: develop/tutorials/articles/120-customizing/10-dynamic-includes/02-top-head-jsp-dynamic-include.markdown [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    	private Portal _portal;
+
+    	private String _postfix;
+
+    	@Reference(target = "(osgi.web.symbolicname=com.liferay.mentions.web)")
+    	private ServletContext _servletContext;
+
+    }
+    
+![Figure 1: The top_head pre key loads additional CSS and HTML resources in the head of the theme.](../../../images/dynamic-include-top-head-pre-mentions-css.png)
```
Comment (jbalsas) [inline develop/tutorials/articles/120-customizing/10-dynamic-includes/02-top-head-jsp-dynamic-include.markdown]: Looks great! Nice idea! I'm just worried a bit about the `liferay-master.wedeploy.io" url... but not sure if there's anything we can do about it.

File: develop/tutorials/articles/120-customizing/10-dynamic-includes/02-top-head-jsp-dynamic-include.markdown [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    		throws IOException {
+
+    		ThemeDisplay themeDisplay = (ThemeDisplay)request.getAttribute(
+    			WebKeys.THEME_DISPLAY);
+
+    		PrintWriter printWriter = response.getWriter();
+
+    		String content = "<link href=\"".concat(themeDisplay.getCDNBaseURL());
+
+    		printWriter.println(content.concat(_postfix));
```
Comment (jbalsas) [inline develop/tutorials/articles/120-customizing/10-dynamic-includes/02-top-head-jsp-dynamic-include.markdown]: Just simplify this. Remove the `activate` method and here, simply have

```java
PrintWriter printWriter = response.getWriter();

String content = "<link href="..." />";

printWriter.println(content);
```

File: develop/tutorials/articles/120-customizing/10-dynamic-includes/02-top-head-jsp-dynamic-include.markdown (hunk 12 above)
Comment (mwilliams2014) [reply develop/tutorials/articles/120-customizing/10-dynamic-includes/02-top-head-jsp-dynamic-include.markdown]: I can change that in the developer console so it reads `http://localhost:8080/` instead.

File: develop/tutorials/articles/120-customizing/10-dynamic-includes/03-top-js-dynamic-include.markdown [hunk 14]
```diff
@@ -0,0 +1,16 @@
+# TOP JS Dynamic Include
+
```
Comment (mwilliams2014) [inline develop/tutorials/articles/120-customizing/10-dynamic-includes/03-top-js-dynamic-include.markdown]: @jbalsas can you provide an example for this? There is only one use in portal and it's a bit long and hard to follow: https://github.com/liferay/liferay-portal/blob/7.1.x/modules/apps/foundation/frontend-js/frontend-js-top-head-extender/src/main/java/com/liferay/frontend/js/top/head/extender/internal/servlet/taglib/TopHeadDynamicInclude.java

File: develop/tutorials/articles/120-customizing/10-dynamic-includes/02-top-head-jsp-dynamic-include.markdown (hunk 12 above)
Comment (jbalsas) [reply develop/tutorials/articles/120-customizing/10-dynamic-includes/02-top-head-jsp-dynamic-include.markdown]: nah, that's fine for now

File: develop/tutorials/articles/120-customizing/10-dynamic-includes/03-top-js-dynamic-include.markdown (hunk 14 above)
Comment (jbalsas) [reply develop/tutorials/articles/120-customizing/10-dynamic-includes/03-top-js-dynamic-include.markdown]: I will, just give me some time :)

File: develop/tutorials/articles/120-customizing/10-dynamic-includes/02-top-head-jsp-dynamic-include.markdown (hunk 12 above)
Comment (mwilliams2014) [reply develop/tutorials/articles/120-customizing/10-dynamic-includes/02-top-head-jsp-dynamic-include.markdown]: k

File: develop/tutorials/articles/120-customizing/10-dynamic-includes/03-top-js-dynamic-include.markdown (hunk 14 above)
Comment (jbalsas) [reply develop/tutorials/articles/120-customizing/10-dynamic-includes/03-top-js-dynamic-include.markdown]: Hey, @mwilliams2014, this should be similar to the wysiwyg resources dynamic include, that lets you add resources to the top of the page, so expect to inject `<script>` tags with, for example, `yourfile.js` that you need present in the head of the page.

This can be a tracking script (analytics) or some module or library you always want present

**Comment by jbalsas** [comment]
Hey @mwilliams2014, I've added another example, and while we're doing this, I think we could start compiling the list of extension points in the intro page.

I'm sure they will likely show as related articles, but what if we also add a regular table with the introduction text so it's easier to digest?

Something like:

Extension Point | Purpose
--- | --- 
top_head | Add things blabbla...
wysiwg | Control the editors...

File: develop/tutorials/articles/120-customizing/10-dynamic-includes/03-top-js-dynamic-include.markdown (hunk 14 above)
Comment (mwilliams2014) [reply develop/tutorials/articles/120-customizing/10-dynamic-includes/03-top-js-dynamic-include.markdown]: ok. So, I can just add another printwriter example that adds the script. I'll make the changes and push them here. Thanks!

**Comment by mwilliams2014** [comment]
@jbalsas updated with your suggestions. Please review the changes and let me know if anything else needs updated before I send forward. Thanks!

File: develop/tutorials/articles/120-customizing/10-dynamic-includes/02-top-head-jsp-dynamic-include.markdown [hunk 15]
```diff
@@ -28,7 +28,7 @@ The example below injects a link into the top of the `top_head.jsp`:
 
     		PrintWriter printWriter = response.getWriter();
 
-    		String content = "<link href="..." />";
+    		String content = "<link href="...mentions.css" />";
```
Comment (jbalsas) [inline develop/tutorials/articles/120-customizing/10-dynamic-includes/02-top-head-jsp-dynamic-include.markdown]: What's this `...`? I think we need a proper path, even if just starting with `/` :)

**Comment by jbalsas** [comment]
Hey @mwilliams2014, looks good, just a weird `css` src in there starting in `...`. Fix that path and send this forward! 👍

File: develop/tutorials/articles/120-customizing/10-dynamic-includes/02-top-head-jsp-dynamic-include.markdown (hunk 15 above)
Comment (mwilliams2014) [reply develop/tutorials/articles/120-customizing/10-dynamic-includes/02-top-head-jsp-dynamic-include.markdown]: Oh, I took your earlier comment literal and put the ... instead of the full path :-D I'll update it.

**Comment by mwilliams2014** [comment]
Will do. Thanks!
